Add a policy for inclusion in the prelude (#66)
rust-lang/std-dev-guide: src/policy/prelude.md

Policy for inclusion in the prelude
===================================

The Rust standard library provides a "prelude": items that Rust programs can
use without having to import anything. For instance, the Rust prelude includes
[`Vec`](https://doc.rust-lang.org/std/vec/struct.Vec.html), so programs can
write `Vec::with_capacity(4)` without having to first import `Vec` from the
standard library.

Each edition of Rust has its own prelude, so that new editions of Rust may add
items without making those items available to all code in older editions of
Rust.

This policy sets out the requirements for what items we add to the prelude, and
whether we add those items to the prelude for a future edition or to the common
prelude for all Rust editions.

When to use editions
--------------------

Adding a trait to the prelude can break existing Rust code, by creating name
resolution ambiguity where none previously existed. While introducing
resolution ambiguities may be "permitted breakage", doing so is quite
disruptive, and we want to avoid doing so. Thus, we generally never add a trait
to the common prelude; we only add traits to the prelude for a new edition of
Rust.

> Adding items *other* than traits to the prelude will never produce conflicts or
> other compatibility issues with existing code, since we allow shadowing and

Reviewer: ```suggestion
Adding items *other* than traits to the prelude introduces less, but not zero,
name resolution conflicts with existing code, since we allow shadowing and
```

Author: I've fixed this with a substantial expansion to the documentation.

[Some kinds of additions to the prelude can cause breakage in existing
code](../breaking-changes/prelude.md), so we must either avoid making such
changes entirely or check carefully with crater before making them.

Adding a trait to the prelude can break existing Rust code, by creating name
resolution ambiguity where none previously existed. While introducing
resolution ambiguities may be "permitted breakage", doing so is quite
disruptive, and we want to avoid doing so. Thus, we generally never add a trait
to the common prelude; we only add traits to the preludes for new editions of
Rust.

Likewise, adding a macro to the prelude can break existing code, so we
generally can't add macros to the common prelude, and can only add macros to
the preludes for new editions of Rust.

Finally, in some cases adding a type to the prelude can break existing code.
The cases in which this arises are narrower, but still require at a minimum
checking a crater run.

Adding items other than traits, macros, or types to the prelude will never
produce conflicts or other compatibility issues with existing code, since we
allow shadowing and give other sources of names priority over names from the

> prelude. Thus, if we choose to add another kind of item to the prelude, we

Reviewer 2: ```suggestion
Adding items other than traits, macros, or types to the prelude will never
produce conflicts or other compatibility issues with existing code, since we
allow shadowing and give other sources of names priority over names from the
prelude. This [includes](https://doc.rust-lang.org/reference/names/namespaces.html#r-names.namespaces.kinds) statics, constants and functions. Thus, if we choose to add another kind of item to the prelude, we
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/std-dev-guide/66/ddb83c8ccaaf40a2bba96fa9d3366503650169e6..efac767dfc5ef30e11baf82c349cd1c3b1d8e930)*

Policy for inclusion in the prelude
===================================

The Rust standard library provides a "prelude": items that Rust programs can
use without having to import anything. For instance, the Rust prelude includes
[`Vec`](https://doc.rust-lang.org/std/vec/struct.Vec.html), so programs can
write `Vec::with_capacity(4)` without having to first import `Vec` from the
standard library.

Each edition of Rust has its own prelude, so that new editions of Rust may add
items without making those items available to all code in older editions of
Rust.

This policy sets out the requirements for what items we add to the prelude, and
whether we add those items to the prelude for a future edition or to the common
prelude for all Rust editions.

This policy provides guidelines, and documents what we can potentially add; it

> does not attempt to mandate what we must add.

Reviewer 2: This might not fix the concern @dtolnay raised during the meeting? "what we **can** add" sounds like something that can enable negative reasoning?

Stronger weakening (heh) of wording:

```suggestion
This policy is advisory, and documents what to consider when approving additions; it
does not attempt to mandate what we must add.
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/std-dev-guide/66/ddb83c8ccaaf40a2bba96fa9d3366503650169e6..c3bd1534344e20c4d54643e1ad14e482635cf6cf)*

Author: "can *potentially* add", it's addressing what would cause breakage. But I've reworded it to be even more explicit now.
